File: src/levels/L12-chain/chain.rb
# Cada eslabón atiende lo suyo o lo pasa al siguiente. Nadie conoce la cadena completa.
class Handler
  def initialize(next_handler = nil) = @next = next_handler

  def handle(request)
    can_handle?(request) ? @refunds.pay(request) : pass(request)
  end

  def pass(request) = @next ? @next.handle(request) : @refunds.decline(request)
end

# region: Cashier#handle
class Cashier < Handler
  def can_handle?(r) = r.amount < 100
end
# endregion

# region: Manager#handle
class Manager < Handler
  def can_handle?(r) = r.amount < 500
end
# endregion

# region: Owner#handle
class Owner < Handler
  def can_handle?(r) = r.amount < 2000
end
# endregion

# region: wiring
chain = Cashier.new(Manager.new(Owner.new))
# endregion

File: src/levels/L12-chain/ts/chain.ts
// Cada eslabón atiende lo suyo o lo pasa al siguiente. Nadie conoce la cadena completa.
abstract class Handler {
  private refunds = new Refunds()

  constructor(private next?: Handler) {}

  abstract canHandle(r: RefundRequest): boolean

  handle(request: RefundRequest) {
    this.canHandle(request) ? this.refunds.pay(request) : this.pass(request)
  }

  pass(request: RefundRequest) { this.next ? this.next.handle(request) : this.refunds.decline(request) }
}

// region: Cashier#handle
class Cashier extends Handler {
  canHandle(r: RefundRequest) { return r.amount < 100 }
}
// endregion

// region: Manager#handle
class Manager extends Handler {
  canHandle(r: RefundRequest) { return r.amount < 500 }
}
// endregion

// region: Owner#handle
class Owner extends Handler {
  canHandle(r: RefundRequest) { return r.amount < 2000 }
}
// endregion

// region: wiring
const chain = new Cashier(new Manager(new Owner()))
// endregion
